backend/tomato/host.py

```diff
@@ -1029,41 +1079,96 @@ def select(site=None, elementTypes=None, connectionTypes=None, networkKinds=None
-	for h in hosts:
-		if h in hostPrefs:
-			prefs[h] += hostPrefs[h]
-		if h.site in sitePrefs:
-			prefs[h] += sitePrefs[h.site]
-	#STEP 4: select the best host
-	hosts.sort(key=lambda h: prefs[h], reverse=True)
-	logging.logMessage("select", category="host", result=hosts[0].name,
+	prefs = dict([(h, getHostValue(h, site, elementTypes, connectionTypes, networkKinds, hostPrefs, sitePrefs)) for h in hosts])
+	return hosts, prefs
+
+def getBestHost(site=None, elementTypes=None, connectionTypes=None,networkKinds=None, hostPrefs=None, sitePrefs=None):
+	hosts_all, prefs = getHostList(site, elementTypes, connectionTypes,networkKinds, hostPrefs, sitePrefs)
+	cand = checkForHostDeactivation()
+	hosts = []
+	for host in hosts_all:
+		if host not in cand:
+			hosts.append(host)
+	if hosts != []:
+		hosts.sort(key=	lambda h: prefs[h], reverse=True)
+	else:
+		hosts = hosts_all.sort(key=	lambda h: prefs[h], reverse=True)
+	return hosts[0], prefs[hosts[0]]
+	
+def checkForHostDeactivation():
+	hosts, prefs = getHostList()
+	candidates = []
+	candidates_prefs = []
+	
+		
+
+	import statistics
+	
+	avg = []
+	for h in getHostList():
+		avg.append(h.getLoad())
+		
+	avg = statistics.mean(avg)
+
+	if avg < AVG_MINIMUM:
+		for host_ in hosts:
+			if host_.detachable:
+				host_elements = HostElement.objects.filter(host = host_)
+				n = 0
+				for el in host_elements:
+					if el.state in ["started"]:
+						n+=1
+				if n == 0:
+					candidates.append(host_)
+					candidates_prefs.append((host_, prefs[host_]))
+		candidates.sort(key=lambda h: prefs[h], reverse=True)
+		
+	return candidates
+	
+def reallocate():
+	
+	import elements
+	#needs to be redefined at a better place
+	THRESHOLD = 20
+	#Walk through all elements and think about reallocating them.
+	for el in  list(elements.getAll()):
+		if el.state in ["started","created"]:
+			continue
+		hostPref, sitePref = el.getLocationPrefs()
+		mainElement = el.mainElement()
+		if mainElement != None and el.element:
+			prevScor = getHostValue(host=mainElement.host,site=mainElement.host.site,elementTypes=mainElement.type,hostPrefs=hostPref,sitePrefs=sitePref)
+			best,bestScor = getBestHost(site=mainElement.host.site,hostPrefs=hostPref,sitePrefs=sitePref)
+		
+			#Compare best host with preference host and migrate to better one if possible
+			if el.element.host != best:
+				if bestScor - prevScor > THRESHOLD:
+					if el.checkMigrate():
+						el.action_migrate(best)
+				
+
+
+def select(site=None, elementTypes=None, connectionTypes=None, networkKinds=None, hostPrefs=None, sitePrefs=None):
+	if not sitePrefs: sitePrefs = {}
+	if not hostPrefs: hostPrefs = {}
+	if not networkKinds: networkKinds = []
+	if not connectionTypes: connectionTypes = []
+	if not elementTypes: elementTypes = []
+	
+	host, pref = getBestHost(site, elementTypes, connectionTypes,networkKinds, hostPrefs, sitePrefs)
+	hosts, prefs = getHostList(site, elementTypes, connectionTypes,networkKinds, hostPrefs, sitePrefs)
+
```

Comment:
This will call ``getHostList()`` over and over again.
  * ``getBestHost`` calls ``getHostList``
  * ``getBestHost`` calls ``checkForHostDeactivation`` which calls ``getHostList``
  * and finally ``getHostList`` is called directly